Use central phi functions instead LST ones

VourMa opened the pull request:
This is the follow up for #142. I have created it as a draft PR, as, to be submitted to cms-sw, we would need to first merge #145 (for proper naming of the `reducePhiRange` functions) and #141 (so that we can have meaningful tests). Having said that, I think we can start discussing and testing this internally.

VourMa commented:
/run all

VourMa commented:
@ariostas Sorry, I forgot, did we move to 15_0? Could you tell me the exact version, so that I can write it on our repo, and then update this PR appropriately?

ariostas commented:
@VourMa I set it up so that now it always uses the latest release, so it's using 15_0_0_pre2. For some reason `AlpakaMath` didn't make it into that release, so I'll just make the CI check out that package so that this PR works

VourMa commented:
> @VourMa I set it up so that now it always uses the latest release, so it's using 15_0_0_pre2. For some reason `AlpakaMath` didn't make it into that release, so I'll just make the CI check out that package so that this PR works

Oops, I didn't think of that. Feel free to just run the tests when you sort it out. Thank you for taking care of it!

ariostas commented:
It should work now.

/run all

VourMa commented:
> https://github.com/SegmentLinking/cmssw/blob/eec20cb881a6d78d6fc5b68ba10346edb417ea1c/RecoTracker/LSTCore/src/alpaka/NeuralNetwork.h#L41-L48
> 
> Can you remove this one in the inference code I added as well? Should probably check that it doesn't affect the performance plots. I tried using another implementation of the delta phi function and it gave some weird results, not sure if it was just a bug in my old code.

Sure, I replaced it in the new version I pushed. Let's see if it works out.

VourMa commented:
/run all

GNiendorf commented:
Looks good, thanks.

VourMa commented:
Now that cms-sw/cmssw#47154 is merged and given that I have followed up on the comments of this PR and the tests have passed, I am rebasing and making the PR to cms-sw.

VourMa commented:
/run all

VourMa commented:
This is all because of cms-sw/cmssw#47119, so LST cannot run in master, unless we update (but then we won't be able to rely on a pre-release for running - the next pre-release is due on February 4th, so quite a few days away). We need to discuss how to proceed.

slava77 commented:
> This is all because of [cms-sw#47119](https://github.com/cms-sw/cmssw/pull/47119), so LST cannot run in master, unless we update (but then we won't be able to rely on a pre-release for running - the next pre-release is due on February 4th, so quite a few days away). We need to discuss how to proceed.

I think that as we did once, we should move the CI to a recent IB, the alpaka version update entered in CMSSW_15_0_X_2025-01-21-2300. Our master now is from Jan 24. 
IIUC, the CI starts with setting up 15_0_0_pre2; so, we depend on its alpaka version in the CI.

slava77 commented:
@ariostas 
it may be useful to have a way to pass the CMSSW version to the /run command
This would imply that the version is as recent as a known default release
Do we still depend in the setup.sh or similar on explicitly decoded paths of needed externals?

ariostas commented:
> it may be useful to have a way to pass the CMSSW version to the /run command
> This would imply that the version is at least as recent as a known default release
> Do we still depend in the setup.sh or similar on explicitly decoded paths of needed externals?

Currently it uses the latest release (bypassing whatever is in `setup.sh`). I'll changing it to use the latest IB release, since those get updated pretty often. That should solve things, right?

ariostas commented:
Should work now

/run all

slava77 commented:
> Currently it uses the latest release (bypassing whatever is in `setup.sh`). I'll changing it to use the latest IB release, since those get updated pretty often. That should solve things, right?

Yes, selecting the latest should work most of the time.
Occasionally there is a broken IB, but that's usually solved in one cycle (12hrs or less)

slava77 commented:
> There was a problem while building and running in standalone mode. The logs can be found [here](https://github.com/SegmentLinking/cmssw/actions/runs/13032903148).

apparently we are missing the complexity of patch build IBs; some paths needs updating

slava77 commented:
> apparently we are missing the complexity of patch build IBs; some paths needs updating

CMSSW_FULL_RELEASE_BASE

VourMa commented:
Thanks for all the updates, @ariostas! I will fix the conflict, rerun the tests and make the PR to cms-sw.

VourMa commented:
/run all

VourMa commented:
> There was a problem while building and running with CMSSW. The logs can be found [here](https://github.com/SegmentLinking/cmssw/actions/runs/13037417847).

The error seems to be:

> ModuleNotFoundError: No module named 'HLTrigger.Configuration.HLTrigger_EventContent_cff'

Glitch? Because I can see the file there in the master:
![image](https://github.com/user-attachments/assets/5f445fd4-7332-45e7-b28c-7712a99475c6)

ariostas commented:
> Glitch? Because I can see the file there in the master

I'm not sure. https://github.com/SegmentLinking/cmssw/pull/148 failed in the same way

GNiendorf commented:
Do you have to do `git cms-addpkg HLTrigger/Configuration`? Similar to how you told me in my PR to get it to run locally?

ariostas commented:
> Do you have to do `git cms-addpkg HLTrigger/Configuration`? Similar to how you told me in my PR to get it to run locally?

The release already includes that file, so it shouldn't be necessary. I'll look into it

slava77 commented:
> > Do you have to do `git cms-addpkg HLTrigger/Configuration`? Similar to how you told me in my PR to get it to run locally?
> 
> The release already includes that file, so it shouldn't be necessary. I'll look into it

I tried to follow the steps from run.sh on cgpu-1 and the cmsDriver command ran OK.
Oddly enough, the cmssw test worked before the update to 23-1100 in https://github.com/SegmentLinking/cmssw/pull/146#issuecomment-2621974039 as seen in the [job outputs](https://github.com/SegmentLinking/cmssw/actions/runs/13032903148/job/36356148150)

slava77 commented:
> I tried to follow the steps from run.sh on cgpu-1 and the cmsDriver command ran OK.

actually, after I followed more literally, it indeed breaks

slava77 commented:
/run cmssw

slava77 commented:
> > I tried to follow the steps from run.sh on cgpu-1 and the cmsDriver command ran OK.
> 
> actually, after I followed more literally, it indeed breaks

it looks like something is glitchy

slava77 commented:
I think the problem starts at least as early as with 
```
------- copying files from src/HLTrigger/Configuration/scripts -------
```

GNiendorf commented:
> I think the problem starts at least as early as with
> 
> ```
> ------- copying files from src/HLTrigger/Configuration/scripts -------
> ```

Are you saying this because it's empty (ie there's no files copied below that)? I am able to reproduce this locally. If I don't include `git cms-addpkg HLTrigger/Configuration` then it is empty. If I do include it then I get the following:

```
------- copying files from src/HLTrigger/Configuration/scripts -------
>> copied edmPluginCoverage
>> copied hltCheckPrescaleModules
>> copied hltCompareResults
>> copied hltConfigFromDB
>> copied hltDumpStream
>> copied hltFindDuplicates
>> copied hltGetConfiguration
>> copied hltIntegrationTests
>> copied hltListPaths
>> copied hltPhase2UpgradeIntegrationTests
>> copied hltPrintMenuVersions
```

slava77 commented:
> Oh but then it just goes to the next missing one:

I have a fix proposed in https://github.com/SegmentLinking/TrackLooper-actions/pull/19

slava77 commented:
> > Oh but then it just goes to the next missing one:
> 
> I have a fix proposed in [SegmentLinking/TrackLooper-actions#19](https://github.com/SegmentLinking/TrackLooper-actions/pull/19)

I'm not sure how our setup worked before; the apparent symptoms are that a bunch of directories with only `.gitignore` file (I counted 17 packages) are pulled to the current directory: scram thinks that these are cleaned up packages and diligently doesn't allow to pick up files available in the base release.

ariostas commented:
Thank you, Slava! Let's see if it works now.

/run cmssw

VourMa commented:
Thank you all for making it work! Submitting this to cms-sw now.

slava77 commented on RecoTracker/LSTCore/src/alpaka/Kernels.h:
```diff
@@ -162,7 +162,7 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
             float phi2 = __H2F(quintuplets.phi()[jx]);
             float dEta = alpaka::math::abs(acc, eta1 - eta2);
-            float dPhi = calculate_dPhi(phi1, phi2);
+            float dPhi = reco::deltaPhi(phi1, phi2);
```
@VourMa 
remind me please the rationale of using `reco::deltaPhi(T phi1, T phi2)` vs `cms::alpakatools::deltaPhi(TAcc const& acc, T phi1, T phi2)`

VourMa replied:
Closely following what was there before. That said, I am not sure why we ended up with `calculate_dPhi(phi1, phi2)` and not something `alpaka` related.

slava77 replied:
uhm, but isn't the implementation different (the old was a single 2π shift vs new using somewhat pseudo-constexpr reducedRange), 
unless by "closely" you mean "it didn't have acc" before

VourMa replied:
About following closely, it is the latter.
About your first point, my understanding is that using `reduceRange` is safer, as it accounts for multiple "wraparounds" of φ, but let me know if I didn't get it right.

slava77 replied:
I'd aim to use the same code that defines dphi from x,y "directly" and after computing phi. My concern is that the computation path x,y -> phi  followed by a call to dPhi for a pair of items is different from computing it directly from a pair of two `x,y`s in the same parts of the code base.  ... also I'm still bothered by pseudo-constexpr in the reco case as was discussed with Matti.

slava77 replied:
> I don't see an inconsistency, at least not in this part of the code, as we do not compute φ from x, y

that's not what I see: `quintuplets.phi()` defines the phi1,2 here, which in turn eventually comes from `hits.phis()[ihit] = phi(acc, ihit_x, ihit_y)`

slava77 replied:
> but I understand that you would want us to apply `cms::alpakatools::deltaPhi(TAcc const& acc, T x1, T y1, T x2, T y2)` and not `cms::alpakatools::deltaPhi(TAcc const& acc, T phi1, T phi2)`?

no; if we have to cache phi, as in this case of T5s or MDs, they should be cached. I mean, eventually, that `cms::alpakatools::deltaPhi` should be preferred, especially if the inputs are likely also computed with `cms::alpakatools::` or accelerator-dependent code

VourMa replied:
Ok, my bad, I thought it was pT5s we were looking at.

Do you want to change this? If so, do you want to do it in this PR?

VourMa replied:
> no; if we have to cache phi, as in this case of T5s or MDs, they should be cached. I mean, eventually, that `cms::alpakatools::deltaPhi` should be preferred, especially if the inputs are likely also computed with `cms::alpakatools::` or accelerator-dependent code

Ok. Would you like to make a comment in the cms-sw PR to mention that we'd better use the alpaka version, and then I can follow up on that?

slava77 replied:
> Do you want to change this? If so, do you want to do it in this PR?

it's better in this PR

slava77 replied:
> Ok. Would you like to make a comment in the cms-sw PR to mention that we'd better use the alpaka version, and then I can follow up on that?

I stayed away from bringing this up in the cms-sw PR due to the review availability issues for heterogeneous, where weeks long delays are typical

VourMa replied:
Ok, then I will force push the `reco::deltaPhi(T phi1, T phi2)` -> `cms::alpakatools::deltaPhi(TAcc const& acc, T phi1, T phi2)` change a bit later today.
